src/orchestrator.ts

```diff
@@ -200,40 +231,214 @@ export class Orchestrator implements Callbacks {
+            throw (
+                'Failed to find correct source runner with uri ' +
+                streamIdentify.runner
+            )
+        }
+
+        const targetInstantiator =
+            this.channelToInstantiator[streamIdentify.channel]
+
+        // If the instantiator exists, it will send a processed message
+        // When this happens, the onMessageProcessedCb will be called
+        // If the instantiator does not exist, it cannot send a processed message
+        // So the onMessageProcessedCb should be called when the stream is closed on the writer's side
+        if (targetInstantiator) {
+            this.runningMessages[globalSequenceNumber] =
+                sourceRunner.onMessageProcessedCb(
+                    streamIdentify.localSequenceNumber,
+                    streamIdentify.channel,
+                )
+
+            const readerConnected = new Promise<ReceivingStream>((res) => {
+                this.waitForConnectingReceivingStream[globalSequenceNumber] =
+                    res
+            })
+
+            targetInstantiator.streamMessage({
+                channel: streamIdentify.channel,
+                globalSequenceNumber,
+            })
+
+            const receivingStream = await readerConnected
+            const writeToReceiver = promisify(
+                receivingStream.write.bind(receivingStream),
+            )
+            const log = this.logLevels[streamIdentify.channel]
+
+            sendingStream.on('data', async (chunk) => {
+                if (chunk.data !== undefined) {
+                    if (log) {
+                        log(decoder.decode(chunk.data!.data))
+                    }
+                    await writeToReceiver(chunk.data)
+                }
+            })
+
+            sendingStream.on('end', () => {
+                try {
+                    receivingStream.end()
+                } catch (ex) {
+                    if (ex instanceof Error) {
+                        console.error(ex.stack)
+                    }
+                }
+            })
+
+            receivingStream.on('data', async (d) => {
+                try {
+                    await writeToSender(d)
+                } catch (ex) {
+                    console.log('Error happend', ex)
+                }
+            })
+
+            writeToSender({ streamSequenceNumber: globalSequenceNumber })
+        } else {
+            this.logger.error(
+                `Receiving stream message for channel ${streamIdentify.channel} without a connected reader`,
+            )
+            const onEnd = sourceRunner.onMessageProcessedCb(
+                streamIdentify.localSequenceNumber,
+                streamIdentify.channel,
+            )
+
+            let chunkCount = 0
+            writeToSender({ streamSequenceNumber: chunkCount })
+            sendingStream.on('data', () => {
+                writeToSender({ streamSequenceNumber: chunkCount++ })
+            })
```

Comment:
I'm wondering if this wouldn't send `writeToSender({ streamSequenceNumber: 0 })` twice? As on first data, `chunkCount++` will resolve to `0` (`++chunkCount` would resolve to 1).

Reply:
Yess correct, doesn't really matter in the current implemented runners, but we should make it correct